File: julia/Euler/src/bin/p0042.jl

# project euler: problem 42

module Prob0042

import ..Util: is_triangular

function solve_0042(fname::String = "p042_words.txt")
    worth(word) = sum(Int(x) - Int('A') + 1 for x in word)
    data = readline(joinpath((@__DIR__), "../../assets", fname))

    count(is_triangular, worth.(split(replace(data, "\"" => ""), ",")))
end

end #module

using .Prob0042: solve_0042
export solve_0042

File: clojure/src/project_euler/sol/p0042.clj
(ns project-euler.sol.p0042
  (:require
   [clojure.string :as str]
   [project-euler.lib.math :as my-math]
   [project-euler.lib.util :as util]))

(defn- parse-data
  [data]
  (-> (first data)
      (str/replace #"\"" "")
      (str/split #",")))

(defn- worth
  [word]
  (->> (map #(- (int %) (dec (int \A))) word)
       (reduce +)))

(defn solve
  ([]
   (solve "0042_words.txt"))
  ([fname]
   (->> (parse-data (util/read-data fname))
        (filter #(my-math/triangular? (worth %)))
        (count))))
